### PR TITLE
Refactor require-tree and split it into different modules

### PR DESCRIPTION
**Pull request by vadimdemedes**
Refs #5492 
### Changes
- Remove `core/server/require-tree.js` completely
- Add `validate-themes` module (https://github.com/vdemedes/Ghost/blob/require-tree-refactor/core/server/utils/validate-themes.js) to scan `content/themes` and return errors/warnings
- Add `read-directory` module (https://github.com/vdemedes/Ghost/blob/require-tree-refactor/core/server/utils/read-directory.js) to recursively read directory. More description below.
- Add `parse-package-json` module (https://github.com/vdemedes/Ghost/blob/require-tree-refactor/core/server/utils/parse-package-json.js) to parse json and validate package.json requirements (mostly a rewrite of what was there before)
- Rewrite `core/server/models/index.js` to manually require models, results in a cleaner code
### New module for reading directories

So, `require-tree` is now replaced with `read-directory` module. The main functionality is the same, it recursively reads directory. When it sees a `package.json` file, it parses the file using `parse-package-json`. The new module also follows symlinks.

There's a new feature - setting file names to ignore. By default, `read-directory` ignores `node_modules` and `bower_components`.

``` js
var readDirectory = require('./core/server/utils/read-directory');

readDirectory(config.paths.themePath, { ignore: ['casper'] }).then(function (tree) {
  // all files/folders named 'casper' are ignored
});
```

Let's discuss!

### CODE REVIEW

File: core/server/models/index.js [hunk 1]
```diff
@@ -1,61 +1,71 @@
...
+            return Promise.all(_.map(tags.toJSON(), function mapper(tag) {
+                return self.Tag.destroy({id: tag.id});
             }));
-        }).then(function () {
-            return self.Tag.findPage({limit: 'all'}).then(function then(data) {
-                return Promise.all(_.map(data.tags, function mapper(tag) {
-                    return self.Tag.destroy({id: tag.id});
-                }));
-            });
         });
```
Comment (ErisDS) [inline core/server/models/index.js]: This part looks like the rebase might have gone wrong - `.findAll` has been removed :wink:

Comment (vdemedes) [inline core/server/models/index.js]: Good catch, fix is coming ;)

File: core/server/utils/parse-package-json.js [hunk 2]
```diff
@@ -0,0 +1,56 @@
+'use strict';
```
Comment (ErisDS) [inline core/server/utils/parse-package-json.js]: afaik, this is the default and therefore not required (we removed these from the core code many moons ago)

**Comment by ErisDS** [comment]
I just tested this out locally on my dev env where I have a few themes, a few zip files, and some random files in my `content/themes` directory. I am seeing slightly different behaviour between this PR and master:

On master, I get 2 'missing package.json' file warnings, one for each of 2 folders which are themes which genuinely are missing the package.json. 

With this PR, I get 7 of those warnings, including one for `.DS_Store` and ones for the zip files etc. 

Interestingly, the list shown in `settings/general` for what constitutes a 'theme' is the same for both and incorrectly contains the .zip and other files. Neither list contains `.DS_Store`.

This is likely a side effect of the way the code has been refactored to unify the logic, meaning both the reading and the validation are occurring on the same set of things. The behaviour should be updated to ignore files and only read directories (inc symlinks), but that can be done separately to this PR. 

With the exception of the incorrectly rebased `findAll` code, I think this is good to merge :+1:

File: core/server/models/index.js [hunk 3]
```diff
@@ -1,61 +1,69 @@
...
+/**
+ * TODO: move to some other place
+ */
+
+// ### deleteAllContent
+// Delete all content from the database (posts, tags, tags_posts)
+exports.deleteAllContent = function deleteAllContent() {
+    var self = exports;
+
+    return self.Post.findPage({limit: 'all'}).then(function then(data) {
```
Comment (ErisDS) [inline core/server/models/index.js]: There is still a rebase error here, this code should look like this: https://github.com/TryGhost/Ghost/blob/master/core/server/models/index.js#L44

Comment (ErisDS) [inline core/server/models/index.js]: This is why the tests are failing :wink:

Comment (vdemedes) [inline core/server/models/index.js]: Ah, totally missed it, the test on travis was not running for more than half an hour, so I forgot about it. Fix submitted, tests are passing!

**Pull request merged**